### ISSUE TITLE
Some installations make permanent changes on the system

### ISSUE THREAD

**Issue by salvaom** [mottosso/rez-scoopz#13]
Some scoop installations like Python 2.7 and 3.7 make permanent changes on the system such as adding the temporal python installation to the `PATH` (py2), changing the registry entry for python2 installation (`HKEY_CURRENT_USER\Software\Python\PythonCore\2.7\InstallPath`) or adding new extensions to `PATHEXT`.

I'm not sure how this can be avoided since the Scoop installation scripts can contain arbitrary commands in the scripts.

**Comment by mottosso** [comment]
The general philosophy of Scoop is that packages should be encapsulated, and Python 2 and 3 actually works well as Rez packages already. Most packages it installs, including Python, aren't running the default installer, but rather unzips the installer and copies files explicitly. This avoids the issue of arbitrary installers running arbitrary commands on your system.

Scoop itself does edit your PATH per default, to append its own encapsulated package. This project snips that in the bud, and instead adds those to the Rez package. If there are other variables modified or added by a Scoop package, then those are added as well; I think already? If not, then Scoop neatly encapsulates those variables in the package definition so they can be ported to Rez.

Scoop isn't like Chocolatey in this sense, which is more like a shortcut to download and run arbitrary installers.

There is #4 which I encountered early on; it would add the temporary paths rez-scoopz uses during an install to your PATH. They won't affect anything, as those dirs are removed on completion, but can be annoying if they are still there.

If you encounter any other changes, this is a good place to submit an issue. Did you not manage to install Python 2 or 3?

**Comment by salvaom** [comment]
I don't recall having any issue with Python 3, and I've seen how they use wix (dark) and lessmsi to extract the installer and directly copy/paste the contents without the installation so that part is behaving as you said and should.

The issue I believe comes from their installation scripts, you can see in Python 2 how it modifies the Windows registry to enable `py.exe` to know about the newly installed versions [here](https://github.com/ScoopInstaller/Versions/blob/master/bucket/python27.json#L35)

Python 3 has the same thing but in it's case it's [wrapped](https://github.com/ScoopInstaller/Main/blob/master/bucket/python.json#L26) around a condition (`$global`) that I'm guessing you take into account.

**Comment by mottosso** [comment]
Yeah, that seems like a flawed Python 2 installer. Python 3 is doing what it's supposed to; `global` is a flag you can optionally pass to Scoop to do just that; install globally. rez-scoopz isn't using that.

With that in mind, it seems reasonable to expect Scoop does sometime modify global state, but isn't supposed to. That is, it's a bug. If it does, I would expect state to get modified globally on the machine where Scoop packages are installed, and converted to Rez packages. But for them to then be stored centrally, and used on another machine where that global state is no more. I would expect those packages to still work as expected.

What I'd really like to do, is snip the ability for Scoop to perform these global actions; by e.g. replacing system calls with no-ops. I haven't looked into this, other than for Scoop-specific features like its auto-update which was replaced. For arbitrary scripts, maybe the best option is to simply snip the whole `post_install` section. Or maybe just the `$create_reg` command, in case there is only a small set of these available.

**Comment by salvaom** [comment]
I was looking into Windows Sandbox as a solution to this but I can't find a command line usage. Software such as WS or Sandboxie could be a solution to this but it feels like way too much.

Probably the `post_install` could solve many issues, but we don't know what software depends on that to properly function.

**Comment by mottosso** [comment]
I tried Sandbox a month or so ago, and it's far from ready yet. It wasn't able to install any `msi` packages, like Python. But yes, conceptually, that could work. In practice though, I expect packages are installed without admin privileges in which case it can't make global changes even if it wanted.
